src/types/Utf8Utf16Convert.cpp

```diff
@@ -383,94 +252,31 @@ void u16state::reset() noexcept
 // Arguments:
 // - in - wstring_view of the UTF-16 string to be converted
 // - out - reference to the resulting UTF-8 string
-// - discardInvalids - if `true` invalid characters are discarded instead of replaced with U+FFFD, default is `false`
 // Return Value:
-// - S_OK          - the conversion succeded without any change of the represented code points
-// - S_FALSE       - the incoming string contained at least one invalid character
+// - S_OK          - the conversion succeded
 // - E_OUTOFMEMORY - the function failed to allocate memory for the resulting string
-// - E_ABORT       - the resulting string length would exceed the max_size and thus, the conversion was aborted before the conversion has been completed
+// - E_ABORT       - the resulting string length would exceed the upper boundary of an int and thus, the conversion was aborted before the conversion has been completed
 // - E_UNEXPECTED  - an unexpected error occurred
-[[nodiscard]] HRESULT u16u8(const std::wstring_view in, std::string& out, bool discardInvalids) noexcept
+[[nodiscard]] HRESULT til::u16u8(const std::wstring_view in, std::string& out) noexcept
 {
-    constexpr const uint32_t unicodeReplacementChar{ 0xFFFDu };
-
     try
     {
-        HRESULT hRes{ S_OK };
         out.clear();
 
         if (in.empty())
         {
-            return hRes;
-        }
-
-        size_t lengthHint{};
-        if (FAILED(SizeTMult(in.length(), gsl::narrow_cast<size_t>(3u), &lengthHint)))
-        {
-            lengthHint = std::max(out.capacity(), in.length());
+            return S_OK;
         }
 
-        out.reserve(lengthHint); // avoid any further re-allocations and copying
+        int lengthIn{};
+        int lengthRequired{};
+        RETURN_HR_IF(E_ABORT, FAILED(SizeTToInt(in.length(), &lengthIn)) || FAILED(IntMult(lengthIn, 3, &lengthRequired)));

```

Comment:
Should we use some of the new chromium safe math stuff here? (see dep/chromium/base/numerics with safe conversions, safe math, etc.)

I haven't figured out how to use it yet, but this might be a good opportunity.

Reply:
Now that #4144 landed we probably should. I have to figure out how to merge it in though 🤔

Reply:
I filed #4290 as follow up.